libcudacxx/.upstream-tests/test/cuda/memcpy_async/memcpy_async_tx.pass.cpp

```diff
@@ -0,0 +1,69 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+//
+// UNSUPPORTED: libcpp-has-no-threads
+// UNSUPPORTED: pre-sm-90
+
+// <cuda/barrier>
+
+#include <cuda/barrier>
+#include <cuda/std/utility> // cuda::std::move
+#include "test_macros.h" // TEST_NV_DIAG_SUPPRESS
+
+// Suppress warning about barrier in shared memory
+TEST_NV_DIAG_SUPPRESS(static_var_with_dynamic_init)
+
+#if defined(__CUDA_MINIMUM_ARCH__) && __CUDA_MINIMUM_ARCH__ < 900
+static_assert(false, "Insufficient CUDA Compute Capability: cuda::device::memcpy_async_tx is not available.");
+#endif // __CUDA_MINIMUM_ARCH__
+
+__device__ alignas(16) int gmem_x[2048];
+
+
+int main(int, char**)
+{
+    NV_IF_TARGET(NV_IS_HOST,(
+        //Required by concurrent_agents_launch to know how many we're launching
+        cuda_thread_count = 512;
+    ));
+
+    NV_DISPATCH_TARGET(
+        NV_IS_DEVICE, (
+            using barrier_t = cuda::barrier<cuda::thread_scope_block>;
+            __shared__ alignas(16) int smem_x[1024];
+            __shared__ barrier_t bar;
```

Comment:
We have a PR open for clang-cuda support and it chokes on that pattern. Could you pelase turn this into:
```suggestion
            __shared__ barrier_t* bar;
```
And then adopt for bar being a pointer?

Reply:
If I apply your suggestion verbatim, then bar will become a pointer to a barrier_t located in shared memory. How do I init that?

Would this work?
```suggestion
            __shared__ uint64_t bar_raw;
            barrier_t & bar = reinterpret_cast<barrier_t &>(bar_raw);
            // ..
                init(&bar, blockDim.x);
```

Reply:
You can work with our initializer:
```suggestion
    __shared__ cuda::barrier<cuda::thread_scope_block>* b;
    shared_memory_selector<cuda::barrier<cuda::thread_scope_block>, constructor_initializer> sel;
    b = sel.construct(2);
```

Reply:
Changed now. I made sure that the pointer itself is not located in shared memory though. That confused me at first.